pkg/logutil/merge_logger.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logutil
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/Godeps/_workspace/src/github.com/coreos/pkg/capnslog"
+)
+
+// line represents a log line that can be printed out
+// through capnslog.PackageLogger.
+type line struct {
+	level capnslog.LogLevel
+	str   string
+}
+
+type status struct {
+	period time.Duration
+
+	mu    sync.Mutex // protect fields below
+	start time.Time  // time to start merging
+	count int        // number of lines from start field
+}
+
+func (s *status) canPrint(now time.Time) bool {
+	return s.period > 0 && s.start.Add(s.period).Before(now)
+}
+
+// MergeLogger merges repeated log lines and prints summary log lines.
+// It doesn't merge log at DEBUG and TRACE level because they should
+// be verbose.
+//
+// For one log line, MergeLogger prints out the line when the line appears
+// at the first time. MergeLogger holds the same log line printed within
+// the merge interval, and prints out summary log line after the interval
+// elapses. It stops merging when the line doesn't appear within the
+// merge interval, or is flushed out.
+// If the merge interval is 0, MergeLogger always merges the line
+// until the line is flushed out.
+type MergeLogger struct {
+	*capnslog.PackageLogger
+
+	mu      sync.Mutex // protect statusm
+	statusm map[line]*status
+}
+
+func NewMergeLogger(logger *capnslog.PackageLogger) *MergeLogger {
+	l := &MergeLogger{
+		PackageLogger: logger,
+		statusm:       make(map[line]*status),
+	}
+	go l.outputLoop()
+	return l
+}
+
+func (l *MergeLogger) Info(mergePeriod time.Duration, entries ...interface{}) {
```

Comment:
The rule will be simple. We print at most 1 same warning per second. We do not print out anything for an inactive peer until it is active again.